plugin/trino-exchange/src/main/java/io/trino/plugin/exchange/s3/S3ExchangeStorage.java

```diff
@@ -0,0 +1,686 @@
+                .withDelimiter(PATH_SEPARATOR);
+        return s3.listObjectsV2(request);
+    }
+
+    private static String keyFromUri(URI uri)
+    {
+        checkArgument(uri.isAbsolute(), "Uri is not absolute: %s", uri);
+        String key = nullToEmpty(uri.getPath());
+        if (key.startsWith(PATH_SEPARATOR)) {
+            key = key.substring(PATH_SEPARATOR.length());
+        }
+        if (key.endsWith(PATH_SEPARATOR)) {
+            key = key.substring(0, key.length() - PATH_SEPARATOR.length());
+        }
+        return key;
+    }
+
+    private static String getBucketName(URI uri)
+    {
+        if (uri.getHost() != null) {
+            return uri.getHost();
+        }
+
+        if (uri.getUserInfo() == null) {
+            return uri.getAuthority();
+        }
+
+        throw new IllegalArgumentException("Unable to determine S3 bucket from URI.");
+    }
+
+    private AmazonS3 createAmazonS3Client(ClientConfiguration clientConfig, String accessKey, String secretKey)
+    {
+        if (isNullOrEmpty(accessKey) || isNullOrEmpty(secretKey)) {
+            throw new IllegalArgumentException("S3 access key and secret key must be configured!");
+        }
+        AWSCredentials awsCredentials = new BasicAWSCredentials(accessKey, secretKey);
+
+        AmazonS3Builder<AmazonS3ClientBuilder, AmazonS3> clientBuilder = AmazonS3Client.builder()
+                .withCredentials(new AWSStaticCredentialsProvider(awsCredentials))
+                .withClientConfiguration(clientConfig);
+
+        boolean regionOrEndpointSet = false;
+
+        if (pinS3ClientToCurrentRegion) {
+            clientBuilder.setRegion(Regions.getCurrentRegion().getName());
+            regionOrEndpointSet = true;
+        }
+
+        if (endPoint != null) {
+            clientBuilder.setEndpointConfiguration(new EndpointConfiguration(endPoint, null));
+            regionOrEndpointSet = true;
+        }
+
+        if (!regionOrEndpointSet) {
+            clientBuilder.withRegion(US_EAST_1);
+            clientBuilder.setForceGlobalBucketAccessEnabled(true);
+        }
+
+        return clientBuilder.build();
+    }
+
+    private static boolean isDirectory(URI uri)
+    {
+        return uri.toString().endsWith(PATH_SEPARATOR);
+    }
+
+    private static class S3ExchangeSliceOutput
+            extends SliceOutput
+    {
+        private static final int INSTANCE_SIZE = ClassLayout.parseClass(S3ExchangeSliceOutput.class).instanceSize();
+
+        private final AmazonS3 s3;
+        private final String bucketName;
+        private final String key;
+        private final Supplier<String> uploadIdFactory;
+        private final ExecutorService uploadExecutor;
+        private final Optional<SSECustomerKey> sseCustomerKey;
+
+        private int currentPartNumber;
+        private byte[] buffer;
+        private int bufferSize;
+
+        private boolean failed;
+        private Optional<String> uploadId = Optional.empty();
```

Comment:
It looks like you need to make this class safe for concurrent access. It is being called out externally and from the threads from `uploadExecutor`

Reply:
No need to because its caller `ExchangeSink` -> `void add(int partitionId, Slice data);` is synchronized

Reply:
There is no concurrent access to `uploadId` as we wait for previous upload to finish in `close` and `flushBuffer`

Reply:
I did not mean it is concurrently modified. But still, the access pattern does not look right. 
You are setting the value of `uploadId` in `uploadPage` which is being run from separate thread (via executor). Then you are reading the value from different thread (e.g. next call to `flushBuffer`). And there is (if I am reading this correctly) no memory barrier between the two threads. So modification can be not visible to the thread which accesses the field later on.

note: I did not check other mutable fields. Maybe the issue is not only about `uploadId`

Reply:
Seems we need to make `uploadId` volatile